src/csharp/Grpc.Core.Api/Metadata.cs

```diff
@@ -345,15 +346,44 @@ internal byte[] GetSerializedValueUnsafe()
             /// Creates a binary value or ascii value metadata entry from data received from the native layer.
             /// We trust C core to give us well-formed data, so we don't perform any checks or defensive copying.
             /// </summary>
-            internal static Entry CreateUnsafe(string key, byte[] valueBytes)
+            internal static unsafe Entry CreateUnsafe(string key, byte* source, int length)
             {
                 if (HasBinaryHeaderSuffix(key))
                 {
-                    return new Entry(key, null, valueBytes);
+                    byte[] arr;
+                    if (length == 0)
+                    {
+                        arr = EmptyBlob;
+                    }
+                    else
+                    {   // create a local copy in a fresh array
+                        arr = new byte[length];
+                        Marshal.Copy(new IntPtr(source), arr, 0, length);
+                    }
+                    return new Entry(key, null, arr);
+                }
+                else
+                {
+                    string s;
+                    if (length == 0)
+                    {
+                        s = "";
+                    }
+                    else
+                    {
+                        int charCount = EncodingASCII.GetCharCount(source, length);
```

Comment:
a more readable alternative might be to use `String.Create` factory method https://docs.microsoft.com/en-us/dotnet/api/system.string.create?view=netstandard-2.1, but it seems that it's only available on nestandard2.1+ and it requires Span, so the approach you proposed sounds good.

Reply:
you've made me think; there *is* actually a way to do this with `Encoding`, except on NET45; I propose we make use of this, using a shim/polyfill only for the NET45 (see new `Grpc.Core.Api.Utils.EncodingExtensions` class) - then how the actual alloc happens is a BCL concern, and can use evils if desired - specifically, it uses `string str = FastAllocateString(length)`, which does an alloc **without wipe**, then over-writes. What is *really* interesting to me here is the implication re transitive dependencies and nuget resolution: a NET472 application will use the NET45 bits, which means it *wouldn't* use this API; what are your thoughts on adding an additional NET46 TFM? (NET46 is when this API becomes available). Or if not NET46, NET48 - i.e. "add an additional .NET Framework TFM that will allow the maximum level of API surface, leaving NET45 as the down-level fallback; nothing in-between" ?